### ISSUE TITLE
Use npm installed linter when available

### ISSUE THREAD

**Issue by tobiabocchi** [mfussenegger/nvim-lint#519]
Hey! thanks for all this work, I have been using your repo for a few days and I really like it!

I am trying to work with a js project which uses a specific version of `standardjs`, I noticed that it nvim-lint was not picking that one but the  one installed [system wide](https://github.com/mfussenegger/nvim-lint/blob/2cf9ad095130755d7d87f1730bcf33c91ee822e4/lua/lint/linters/standardjs.lua#L21) instead..

For formatting I am using conform.nvim which I think handles this correctly through a util method:

[here](https://github.com/stevearc/conform.nvim/blob/3d59cbd01a4b74124c5a6fb23b8cc63e5c2db3d5/lua/conform/formatters/standardjs.lua#L9) is the code that sets up `standardjs` as formatter, to parse the command to use it uses [this](https://github.com/stevearc/conform.nvim/blob/3d59cbd01a4b74124c5a6fb23b8cc63e5c2db3d5/lua/conform/util.lua#L6) util method.

I thought this might come in handy for a few linters in your project and maybe worth including.. I am not very experienced in lua, I mainly just use it to set up neovim.. If you don't have the time or don't think this is a priority I can try adding it myself and open a PR

**Comment by AbelAnaya** [comment]
Hello!

I achieved this behavior by reusing some code of the `eslint` linter configuration in this project.

You can select how a linter behaves by modifying its properties in your configuration file for the plugin. An example for standard:

```lua
local lint = require("lint")
local binary_name = "standard"

local cmd_string = function()
    local local_binary = vim.fn.fnamemodify("./node_modules/.bin/" .. binary_name, ":p")
    return vim.loop.fs_stat(local_binary) and local_binary or binary_name
end

-- Use always local standard
local standardjs = lint.linters.standardjs
standardjs.cmd = cmd_string()
```

Simply add this code to your configuration function and `nvim-lint` will start using `standardjs` from local `node_modules` if possible, if not it will fallback to system wide installation.

I will try to add a PR that integrates the utils method of `conform` and uses it for `standardjs`in that project.

**Comment by mfussenegger** [comment]
A PR that adapts standardjs to use the same pattern as eslint would be welcome.

But I'm starting wo wonder if this use-case wouldn't be in general better solved by adjusting your `$PATH` to something like `$PATH=./node_modules/.bin/:$PATH` - then you've consistend behavior across terminal and nvim-lint

**Comment by casr** [comment]
> But I'm starting wo wonder if this use-case wouldn't be in general better solved by adjusting your `$PATH` to something like `$PATH=./node_modules/.bin/:$PATH` - then you've consistend behavior across terminal and nvim-lint

@mfussenegger I went down this route earlier trying to get ESLint working on a project (and I suspect it's relevant to the standardjs use case here too) but it wasn't viable due the difference between the cwd in Neovim and where ESLint's config file is. For example:

- Neovim opened at `~/src/project`
- project has two sub-projects `./subdir/one` & `./subdir/two` each with their own `node_modules` folder and own ESLint configs.

In this setup, adjusting the path to include `~/src/project/subdir/two/node_modules/.bin` does not help because ESLint fails to discover `~/src/project/subdir/two/eslint.config.js`, as ESLint is launched with `~/src/project` as the cwd and therefore no rules are applied.

I think the path forward might be to use the current file as a starting point and traverse up the tree in a similar way to how lspconfig handles finding a root. Would that be viable? I don't mind taking a stab at a PR if so

**Comment by mfussenegger** [comment]
> Would that be viable?

No, I don't think it is viable to have logic to find root-paths scattered across all linters.

If you don't open nvim in the project directory, you can make sure to call `try_lint` with a cwd argument. See also https://github.com/mfussenegger/nvim-lint/issues/482#issuecomment-1999185606
